File: Randomized_Bates_model.m
function Randomized_Bates_model
close all;clc;
% Randomization of the Bates model.
% 
% The code provided is based on the article 'On Randomization of Affine Diffusion Processes
% with Application to Pricing of Options on VIX and S&P 500' by Lech A. Grzelak, 
% L.A. Grzelak@uu.nl
% 
% @article{grzelakRAnD,
%   title={On Randomization of Affine Diffusion Processes with Application to Pricing of Options on {VIX} and {S&P} 500},
%   author={Grzelak, Lech A.},
%   journal={arXiv:2208.12518},  
%   year = {2022}
% }

K = linspace(60,140,25)/100;

% The number of Quadrature points
N = 9;

% The Bates model parameters
r         = 0.0;
muJ       = -0.1; 
sigmaJ    = 0.06; 
xiP       = 0.08; 
kappa     = 0.5; 
gamma     = 0.5;
vbar      = 0.13;
rho       = -0.7;
T         = 1/12;
v0        = 0.13; 
S0        = 1;
CP        = 'c';

% REFERENCE: Implied Volatilities for the Bates model with constant
% parameters
cf_Bates       = @(u)ChFBates(u, T, kappa,vbar,gamma,rho, v0, r, muJ, sigmaJ, xiP);
Call_COS_Bates = CallPutOptionPriceCOSMthd(cf_Bates,CP,S0,r,T,K',1000,8);
IV_COS_Bates   = ImpliedVols(K,Call_COS_Bates,S0,T,r,0.3);
figure(1)
set(gca, 'ColorOrder', [0 0 0; 0.7 0.1 0.2;0.7 0.3 0.9;0.1 0.2 0.7]); 
hold on;grid on;
plot(log(K),IV_COS_Bates*100,'r','LineWidth',1.5)

%%%%%%%%%%%%%%%% Randomization of vol-vol (gamma) %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
p1Vec = [0.01,0.2,0.1,0.75];
p2Vec = [0.25,5,2,1.75];

% Randomization with Uniform distribution
leg={strcat('$\gamma = ',num2str(gamma),'$')};
idx =1;
for i =1 :length(p1Vec)
    p1 = p1Vec(i);
    p2 = p2Vec(i);
    [x_i,w_i]    = UniformCollocation(p1,p2,N); 
    EX = sum(x_i.*w_i);
    cf_generic    = @(u)ChFBates_RAnD(u, T, kappa, vbar, rho, v0, r, muJ, sigmaJ, xiP, x_i, w_i);
    Call_COS_RAnD = CallPutOptionPriceCOSMthd(cf_generic,CP,S0,r,T,K',1000,8);
    IV_COS_RAnD   = ImpliedVols(K,Call_COS_RAnD,S0,T,r,0.3);
    plot(log(K),IV_COS_RAnD*100)
    idx = idx +1;
    leg{idx}={strcat('$\gamma \sim ',' \mathcal{U}([',num2str(p1),',',num2str(p2),']),','\;\;\;  {\bf{E}}[\gamma]= ',num2str(EX),'$')};
    
end  

legend(string(leg),'interpreter','latex')
title('Implied Volatilities for RAnD Bates Model')
xlabel('Strike, K (log-moneyness)')
ylabel('Implied Volatility, [%]')

% Exact pricing of European Call/Put option with the Black-Scholes model
function value=BS_Call_Option_Price(CP,S_0,K,sigma,tau,r)
% Black-Scholes Call option price
d1    = (log(S_0 ./ K) + (r + 0.5 * sigma^2) * tau) / (sigma * sqrt(tau));
d2    = d1 - sigma * sqrt(tau);
if lower(CP) == 'c' || lower(CP) == 1
    value =normcdf(d1) * S_0 - normcdf(d2) .* K * exp(-r * tau);
elseif lower(CP) == 'p' || lower(CP) == -1
    value =normcdf(-d2) .* K*exp(-r*tau) - normcdf(-d1)*S_0;
end

function [IV]=ImpliedVols(K,C,S0,T,r,initial)
IV=zeros([length(K),1]);
for i = 1:length(K)
    IV(i)=ImpliedVolatility('c',C(i),K(i),T,S0,r,initial);
end

function impliedVol = ImpliedVolatility(CP,marketPrice,K,T,S_0,r,initialVol)
    func = @(sigma) (BS_Call_Option_Price(CP,S_0,K,sigma,T,r) - marketPrice);
    impliedVol = fzero(func,initialVol);

function cf=ChFBates(u, tau, kappa,vBar,gamma,rho, v0, r, muJ, sigmaJ, xiP)
i     = complex(0,1);

% functions D_1 and g
D_1  = sqrt(((kappa -i*rho*gamma.*u).^2+(u.^2+i*u)*gamma^2));
g    = (kappa- i*rho*gamma*u-D_1)./(kappa-i*rho*gamma*u+D_1);    

% complex valued functions A and C
C = (1/gamma^2)*(1-exp(-D_1*tau))./(1-g.*exp(-D_1*tau)).*(kappa-gamma*rho*i*u-D_1);
A = i*u*r*tau + kappa*vBar*tau/gamma^2 * (kappa-gamma*rho*i*u-D_1)-2*kappa*vBar/gamma^2*log((1-g.*exp(-D_1*tau))./(1-g));

% Adjustment for the Bates model
A = A - xiP*i*u*tau*(exp(muJ+1/2*sigmaJ^2)-1) + xiP*tau*(exp(i*u*muJ-1/2*sigmaJ^2*u.^2)-1);

% ChF for the Bates model
cf = exp(A + C * v0);

function CfTotal=ChFBates_RAnD(u, tau, kappa,vBar,rho, v0, r, muJ, sigmaJ, xiP, x_i, w_i)
i     = complex(0,1);

CfTotal = zeros([1,length(u)]);
for k = 1: length(x_i)
    % Get realization for gamma
    gamma = x_i(k);
                
    % functions D_1 and g
    D_1  = sqrt(((kappa -i*rho*gamma.*u).^2+(u.^2+i*u)*gamma^2));
    g    = (kappa- i*rho*gamma*u-D_1)./(kappa-i*rho*gamma*u+D_1);    

    % complex valued functions A and C
    C = (1/gamma^2)*(1-exp(-D_1*tau))./(1-g.*exp(-D_1*tau)).*(kappa-gamma*rho*i*u-D_1);
    A = i*u*r*tau + kappa*vBar*tau/gamma^2 * (kappa-gamma*rho*i*u-D_1)-2*kappa*vBar/gamma^2*log((1-g.*exp(-D_1*tau))./(1-g));

    % Adjustment for the Bates model
    A = A - xiP*i*u*tau*(exp(muJ+1/2*sigmaJ^2)-1) + xiP*tau*(exp(i*u*muJ-1/2*sigmaJ^2*u.^2)-1);

    % ChF for the Bates model
    cf = exp(A + C * v0);
    CfTotal = CfTotal + cf.*w_i(k);
end

function [x_i,w_i]=UniformCollocation(a,b,N)

a_idx =@(k)a.^(0:1:k);
b_idx =@(k)b.^(k:-1:0);

EX = @(k)1./(k+1)*sum(a_idx(k).*b_idx(k));
% Moments for standard nodmal i.e. N(0,1)
for i=1:N+1
    for j=1:N+1
        if (i==1&&j==1)
        M(i,j)=1;
        else
        M(i,j)=EX(i+j-2);
        end
    end
end

%%% Zeros as the diagonal of Matrix B
[x_i,w_i]     = FindCollocationPoints(M);

function [x_i,w_i] = FindCollocationPoints(M)

[N,~] =size(M);
N =N-1;

R=chol(M);

alpha(1)  =  R(1,2);
beta(1)   =  (R(2,2)/R(1,1))^2;
for i=2:N-1
    alpha(i) =R(i,i+1)/R(i,i)-R(i-1,i)/R(i-1,i-1);
    beta(i)  =(R(i+1,i+1)/R(i,i))^2;
end
alpha(N)=R(N,N+1)/R(N,N)-R(N-1,N)/R(N-1,N-1);

%%% Construction of the array and zeros calculation
J=diag(sqrt(beta),-1)+diag(alpha,0)+diag(sqrt(beta),1);
[w_i,B]   = eig(J);

w_i=(w_i(1,:)).^2;
w_i= w_i';

%%% Zeros as the diagonal of Matrix B
x_i     = diag(B);

function value = CallPutOptionPriceCOSMthd(cf,CP,S0,r,tau,K,N,L)
i = complex(0,1);
% cf   - characteristic function as a functon, in the book denoted as \varphi
% CP   - C for call and P for put
% S0   - Initial stock price
% r    - interest rate (constant)
% tau  - time to maturity
% K    - vector of strikes
% N    - Number of expansion terms
% L    - size of truncation domain (typ.:L=8 or L=10)  

x0 = log(S0 ./ K);   

% Truncation domain
a = 0 - L * sqrt(tau); 
b = 0 + L * sqrt(tau);

k = 0:N-1;              % row vector, index for expansion terms
u = k * pi / (b - a);   % ChF arguments

H_k = CallPutCoefficients(CP,a,b,k);
temp    = (cf(u) .* H_k).';
temp(1) = 0.5 * temp(1);     % adjust the first element by 1/2

mat = exp(i * (x0 - a) * u);  % matrix-vector manipulations

% Final output
value = exp(-r * tau) * K .* real(mat * temp);

% Coefficients H_k for the COS method
function H_k = CallPutCoefficients(CP,a,b,k)
    if lower(CP) == 'c' || CP == 1
        c = 0;
        d = b;
        [Chi_k,Psi_k] = Chi_Psi(a,b,c,d,k);
         if a < b && b < 0.0
            H_k = zeros([length(k),1]);
         else
            H_k = 2.0 / (b - a) * (Chi_k - Psi_k);
         end
    elseif lower(CP) == 'p' || CP == -1
        c = a;
        d = 0.0;
        [Chi_k,Psi_k]  = Chi_Psi(a,b,c,d,k);
         H_k = 2.0 / (b - a) * (- Chi_k + Psi_k);       
    end

function [chi_k,psi_k] = Chi_Psi(a,b,c,d,k)
    psi_k        = sin(k * pi * (d - a) / (b - a)) - sin(k * pi * (c - a)/(b - a));
    psi_k(2:end) = psi_k(2:end) * (b - a) ./ (k(2:end) * pi);
    psi_k(1)     = d - c;
    
    chi_k = 1.0 ./ (1.0 + (k * pi / (b - a)).^2); 
    expr1 = cos(k * pi * (d - a)/(b - a)) * exp(d)  - cos(k * pi... 
                  * (c - a) / (b - a)) * exp(c);
    expr2 = k * pi / (b - a) .* sin(k * pi * ...
                        (d - a) / (b - a))   - k * pi / (b - a) .* sin(k... 
                        * pi * (c - a) / (b - a)) * exp(c);
    chi_k = chi_k .* (expr1 + expr2);

File: Randomized_Bates_model.py
# -*- coding: utf-8 -*-
"""
Randomization of the Bates model.

The code provided is based on the article 'On Randomization of Affine Diffusion Processes
with Application to Pricing of Options on VIX and S&P 500' by Lech A. Grzelak, 
L.A. Grzelak@uu.nl

@article{grzelakRAnD,
title = {On randomization of affine diffusion processes with application to pricing of options on VIX and S&P 500},
journal = {Applied Mathematics and Computation},
volume = {508},
pages = {129598},
year = {2026},
issn = {0096-3003},
doi = {https://doi.org/10.1016/j.amc.2025.129598},
url = {https://www.sciencedirect.com/science/article/pii/S0096300325003248},
author = {Lech A. Grzelak},
}

@author: LECH A. GRZELAK
"""
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st
from scipy.linalg import cholesky as chol
import scipy.linalg as linalg

def CallPutOptionPriceCOSMthd(cf,CP,S0,r,tau,K,N,L):
    # cf   - characteristic function as a functon, in the book denoted as \varphi
    # CP   - C for call and P for put
    # S0   - Initial stock price
    # r    - interest rate (constant)
    # tau  - time to maturity
    # K    - list of strikes
    # N    - Number of expansion terms
    # L    - size of truncation domain (typ.:L=8 or L=10)  
        
    # reshape K to a column vector
    K = np.array(K).reshape([len(K),1])
    
    #assigning i=sqrt(-1)
    i = np.complex(0.0,1.0) 
    
    x0 = np.log(S0 / K)   
    
    # truncation domain
    a = 0.0 - L * np.sqrt(tau)
    b = 0.0 + L * np.sqrt(tau)
    
    # sumation from k = 0 to k=N-1
    k = np.linspace(0,N-1,N).reshape([N,1])  
    u = k * np.pi / (b - a);  

    # Determine coefficients for Put Prices  
    H_k = CallPutCoefficients(CP,a,b,k)
       
    mat = np.exp(i * np.outer((x0 - a) , u))

    temp = cf(u) * H_k 
    temp[0] = 0.5 * temp[0]    
    
    value = np.exp(-r * tau) * K * np.real(mat.dot(temp))
         
    return value

""" 
Determine coefficients for Put Prices 
"""
def CallPutCoefficients(CP,a,b,k):
    if str(CP).lower()=="c" or str(CP).lower()=="1":                  
        c = 0.0
        d = b
        coef = Chi_Psi(a,b,c,d,k)
        Chi_k = coef["chi"]
        Psi_k = coef["psi"]
        if a < b and b < 0.0:
            H_k = np.zeros([len(k),1])
        else:
            H_k      = 2.0 / (b - a) * (Chi_k - Psi_k)  
        
    elif str(CP).lower()=="p" or str(CP).lower()=="-1":
        c = a
        d = 0.0
        coef = Chi_Psi(a,b,c,d,k)
        Chi_k = coef["chi"]
        Psi_k = coef["psi"]
        H_k      = 2.0 / (b - a) * (- Chi_k + Psi_k)               
    
    return H_k    

def Chi_Psi(a,b,c,d,k):
    psi = np.sin(k * np.pi * (d - a) / (b - a)) - np.sin(k * np.pi * (c - a)/(b - a))
    psi[1:] = psi[1:] * (b - a) / (k[1:] * np.pi)
    psi[0] = d - c
    
    chi = 1.0 / (1.0 + np.power((k * np.pi / (b - a)) , 2.0)) 
    expr1 = np.cos(k * np.pi * (d - a)/(b - a)) * np.exp(d)  - np.cos(k * np.pi 
                  * (c - a) / (b - a)) * np.exp(c)
    expr2 = k * np.pi / (b - a) * np.sin(k * np.pi * 
                        (d - a) / (b - a))   - k * np.pi / (b - a) * np.sin(k 
                        * np.pi * (c - a) / (b - a)) * np.exp(c)
    chi = chi * (expr1 + expr2)
    
    value = {"chi":chi,"psi":psi }
    return value

def ImpliedVolatility(CP,V_market,S_0,K,tau,r,sigma0):
    error    = 1e10; # initial error
    Nmax = 100
    #Handy lambda expressions
    optPrice = lambda sigma: BS_Call_Option_Price(CP,S_0,K,sigma,tau,r)
    vega= lambda sigma: dV_dsigma(S_0,K,sigma,tau,r)
    
    sigma = sigma0
    # While the difference between the model and the arket price is large
    # follow the iteration
    i =0
    while error>10e-10 and  i <Nmax:
        f         = V_market - optPrice(sigma);
        f_prim    = -vega(sigma);
        sigma_new = sigma - f / f_prim;
    
        error=abs(sigma_new-sigma);
        sigma=sigma_new;
        i=i+1
        #print(i)
    return sigma

# Vega, dV/dsigma
def dV_dsigma(S_0,K,sigma,tau,r):
    #parameters and value of Vega
    d2   = (np.log(S_0 / (K)) + (r - 0.5 * np.power(sigma,2.0)) * tau) / float(sigma * np.sqrt(tau))
    value = K * np.exp(-r * tau) * st.norm.pdf(d2) * np.sqrt(tau)
    return value

def BS_Call_Option_Price(CP,S_0,K,sigma,tau,r):
    #Black-Scholes Call option price
    d1    = (np.log(S_0 / (K)) + (r + 0.5 * np.power(sigma,2.0)) * tau) / (sigma * np.sqrt(tau))
    d2    = d1 - sigma * np.sqrt(tau)
    if str(CP).lower()=="c" or str(CP).lower()=="1":
        value = st.norm.cdf(d1) * S_0 - st.norm.cdf(d2) * K * np.exp(-r * tau)
    elif str(CP).lower()=="p" or str(CP).lower()=="-1":
        value = st.norm.cdf(-d2) * K * np.exp(-r * tau) - st.norm.cdf(-d1)*S_0
    return value

def CollocationUniform(a,b,N):
    moment = lambda n: (b**(n+1)-a**(n+1))/((n+1)*(b-a))

    # Creation of Matrix M, dimension N+1 x N+1
    M = np.zeros([N+1,N+1])    
    for i in range(0,N+1):
        for j in range(0,N+1):
            M[i,j] =moment(i+j)        
    
    # Once the moments are computed use generic code to calculate quadrature pairs
    x_i, w_i = FindCollocation(M)
    
    return x_i, w_i

def FindCollocation(M):
     # Creation of UPPER diagonal matrix, R, dimension N+1 x N+1

    # Since Matrix M also includes the 0 moment we adjust the size
    N = len(M)-1    
    
    R = chol(M)   
      
    # Creation of vector alpha and beta
    alpha = np.zeros([N])
    beta = np.zeros([N-1])
    
    alpha[0] = R[0,1]
    beta[0]  = (R[1,1]/R[0,0])**2.0
    
    for i in range(1,N):
        alpha[i] = R[i,i+1]/R[i,i] - R[i-1,i]/R[i-1,i-1]
    
    for i in range(1,N-1):
        beta[i]  = (R[i+1,i+1]/R[i,i])**2.0 

    # Construction of matrix J
    J = np.diag(np.sqrt(beta),k=-1)+np.diag(alpha,k=0)+np.diag(np.sqrt(beta),k=1);
    
    # computation of the weights
    eigenValues, eigenVectors = linalg.eig(J)
        
    w_i = eigenVectors[0,:]**2.0
    x_i = np.real(eigenValues)
    
    # sorting the arguments
    idx =np.argsort(x_i)
    w_i = w_i[idx]
    x_i = x_i[idx]
    
    return x_i, w_i

def ChFBatesModel(u,r,tau,kappa,gamma,vbar,v0,rho,xiP,muJ,sigmaJ):
    i = np.complex(0.0,1.0)
    D1 = np.sqrt(np.power(kappa-gamma*rho*i*u,2)+(u*u+i*u)*gamma*gamma)
    g  = (kappa-gamma*rho*i*u-D1)/(kappa-gamma*rho*i*u+D1)
    C  = (1.0-np.exp(-D1*tau))/(gamma*gamma*(1.0-g*\
                               np.exp(-D1*tau)))*(kappa-gamma*rho*i*u-D1)
    # Note that we exclude the term -r*tau, as the discounting is performed in the COS method
    AHes= r * i*u *tau + kappa*vbar*tau/gamma/gamma *(kappa-gamma*\
        rho*i*u-D1) - 2*kappa*vbar/gamma/gamma*np.log((1.0-g*np.exp(-D1*tau))/(1.0-g))
    
    A = AHes - xiP * i * u * tau *(np.exp(muJ+0.5*sigmaJ*sigmaJ) - 1.0) + \
            xiP * tau * (np.exp(i*u*muJ - 0.5 * sigmaJ * sigmaJ * u * u) - 1.0)
    
    # Characteristic function for the Heston's model    
    cf = np.exp(A + C*v0)
    return cf 

def CHF_Bates_RAnD_randomized_gamma(u,r,tau,kappa,gamma,vbar,v0,rho,xiP,muJ,sigmaJ,x_i,w_i):
    ChF_output= np.zeros([len(u),1])
    for n in range(len(x_i)):
        gamma = x_i[n]
        cf = ChFBatesModel(u,r,tau,kappa,gamma,vbar,v0,rho,xiP,muJ,sigmaJ)
        ChF_output = ChF_output + w_i[n]*cf
        
    return ChF_output

def CHF_GBM_RAnD(u,T,r,x_i,w_i):
    i = 1j
    ChF_output= np.zeros([len(u),1])
    for n in range(len(x_i)):
        sigma = x_i[n]
        cf = np.exp((r - 0.5 * sigma**2.0) * i * u * T - 0.5 * sigma**2.0 * u**2.0 * T)
        ChF_output = ChF_output + w_i[n]*cf
    return ChF_output


def main():
    
    # Uniform randomization for the Black Scholes model
    a = 0.1
    b = 0.45
    N = 9
        
    S0 = 1
    CP = 'c'
    r      = 0.0
    muJ    = -0.1 
    sigmaJ = 0.06 
    xiP    = 0.08 
    kappa  = 0.5 
    gamma  = 0.5
    vbar   = 0.13
    rho    = -0.7
    T      = 1/12
    v0     = 0.13 
    Nk     = 25
    K      = np.linspace(0.6,1.4,Nk)
    L = 7
    Ncos = 5000
    
    # Benchmark Bates model
    cf = lambda u: ChFBatesModel(u,r,T,kappa,gamma,vbar,v0,rho,xiP,muJ,sigmaJ)
    
    # Call Option price using the COS Method
    CallValue_Reference = CallPutOptionPriceCOSMthd(cf,CP,S0,r,T,K,Ncos,L)
       
    # Plot option prices
    plt.figure(1)
    plt.plot(K,CallValue_Reference,'--k')
    plt.xlabel('strike K')
    plt.ylabel('Call Option price')
    plt.grid()

    # Compute implied volatilities
    IVRAnD = np.zeros([Nk])            
    for (idx, k) in enumerate(K):
        IVRAnD[idx] = ImpliedVolatility(CP,CallValue_Reference[idx],S0,k,T,r,0.35)

    # Plot implied volatilities
    plt.figure(2)
    plt.plot(K,IVRAnD,'--k')
    plt.xlabel('strike K')
    plt.ylabel('Implied Volatilities')
    plt.title('Implied volatility for Randomized Bates model (RAnD Bates)')
    plt.grid()
    
    # RANDOMIZATION of Gamma parameter with uniformly distributed random variable
    p1Vec = [0.01,0.2,0.1,0.75]
    p2Vec = [0.25,5,2,1.75]
    
    label =['reference, gamma='+str(gamma)]
    
    for n in range(0,len(p1Vec)):
        
        a = p1Vec[n]
        b = p2Vec[n]
        # Compute optimal points, here we take sigma^2 to follow uniform
        x_i, w_i = CollocationUniform(a,b,N)
        
        # Characteristic function for randomized Bates model (randomization of Gamma)
        cf_RAnD = lambda u: CHF_Bates_RAnD_randomized_gamma(u,r,T,kappa,gamma,vbar,v0,rho,xiP,muJ,sigmaJ,x_i,w_i)
        
        # Call Option price using the COS Method
        CallValue_RAnD = CallPutOptionPriceCOSMthd(cf_RAnD,CP,S0,r,T,K,Ncos,L)
        
        # Compute implied volatilities
        IVRAnD = np.zeros([Nk])            
        for (idx, k) in enumerate(K):
            IVRAnD[idx] = ImpliedVolatility(CP,CallValue_RAnD[idx],S0,k,T,r,0.35)
        
        label.append('gamma = U([' + str(a) + ','+ str(b)+'])')
        plt.plot(K,IVRAnD)
    
    plt.legend(label)


main()
